src/lib/utils/object-graph.ts

```diff
@@ -0,0 +1,172 @@
+   * `include` and referenced by ID.
+   */
+  function walk(value: unknown): unknown {
+    if (typeof value !== 'object' || value === null) return value
+
+    const entry = seen.get(value)
+    if (entry?.isDuplicate) {
+      if (!(entry.__refId in include)) {
+        // If we're already serializing this __refId (due to circularity), just
+        // return the ref
+        if (serializing.has(entry.__refId)) {
+          return { __refId: entry.__refId }
+        }
+
+        serializing.add(entry.__refId)
+
+        // Recursively walk object or array content
+        let result: unknown
+        if (Array.isArray(value)) {
+          result = value.map(walk)
+        } else {
+          const obj: Record<string, unknown> = {}
+          for (const [key, val] of Object.entries(value)) {
+            obj[key] = walk(val)
+          }
+          result = obj
+        }
+
+        include[entry.__refId] = result
+        serializing.delete(entry.__refId)
+      }
+
+      return { __refId: entry.__refId }
+    }
+
+    // Not a duplicate, serialize normally
+    if (Array.isArray(value)) {
+      return value.map(walk)
+    }
+
+    const output: Record<string, unknown> = {}
+    for (const [key, val] of Object.entries(value)) {
+      output[key] = walk(val)
+    }
+    return output
+  }
+
+  findDuplicates(input)
+  const data = walk(input) as T
+  return { data, include }
+}
+
+/**
+ * Reconstructs the original object from its serialized form, resolving all
+ * `{ __refId }` references. Preserves object identity and circular structure by
+ * caching resolved references.
+ *
+ * NOTE: The return value is type-asserted, so be careful with the type you
+ * assign to it!
+ */
+export function inflateObjectGraph<T>(data: FlattenedGraph<T>): T {
+  const { data: root, include } = data
+  const cache = new Map<string, unknown>() // Used to maintain shared/circular reference identity
+
+  /**
+   * Recursively resolves `{ __refId }` references into actual object instances.
+   * Uses a cache to prevent reprocessing and to rebuild cycles correctly.
+   */
+  function resolve(value: unknown): unknown {
+    if (Array.isArray(value)) {
+      return value.map(resolve)
+    }
+
+    if (typeof value === 'object' && value !== null) {
+      if ('__refId' in value && typeof value.__refId === 'string') {
+        const __refId = (value as Reference).__refId
+
+        // Return from cache if already resolved (handles shared/circular refs)
+        if (cache.has(__refId)) {
+          return cache.get(__refId)
+        }
+
+        const refData = include[__refId]
+        if (refData === undefined) {
+          throw new Error(`Reference id ${__refId} not found in include map`)
+        }
+
+        // Create placeholder to support circular references
+        const shell: Record<string, unknown> = {}
+        cache.set(__refId, shell)
+
+        const resolved = resolve(refData)
+        Object.assign(shell, resolved)
```

Comment:
It would require too much brain power and time right now for me to reason through the sequence of events here, but I trust you and the automated tests. Instinctively this feels right. 🤯